query_optimizer/logical/Logical.hpp

```diff
@@ -64,12 +70,62 @@ class Logical : public OptimizerTree<Logical> {
    */
   virtual std::vector<expressions::AttributeReferencePtr> getReferencedAttributes() const = 0;
 
+  /**
+   * @brief Make a copy of this logical node with input expressions replaced by
+   *        input_expression.
+   *
+   * @return A copy of this logical node with the new input expressions.
+   */
+  virtual LogicalPtr copyWithNewInputExpressions(
+      const std::vector<expressions::ExpressionPtr> &input_expressions) const {
+    // Default implementation: return a shared reference to "this" when there is
+    // no change.
+    DCHECK(input_expressions.empty());
+    return shared_from_this();
+  }
+
+  /**
+   * @return The input expressions in this Logical.
+   */
+  const std::vector<expressions::ExpressionPtr>& input_expressions() const {
+    return input_expressions_;
+  }
+
+  /**
+   * @return The input attributes, which are the union set of the output attributes of
+   *         all children.
+   */
+  std::vector<expressions::AttributeReferencePtr> getInputAttributes() const {
+    std::vector<expressions::AttributeReferencePtr> input_attributes;
+    for (const LogicalPtr &child : children()) {
+      const std::vector<expressions::AttributeReferencePtr> child_output_attributes =
```

Comment:
`A a(f())` is equivalent to `A a = f()` because here it is construction instead of assignment. I.e.
```
A a = f();  // construction
```
```
A a;
a = f();  // assignment
```
You can check the examples in
http://en.cppreference.com/w/cpp/language/copy_constructor
and 
http://en.cppreference.com/w/cpp/language/copy_assignment